src/sentry/api/serializers/rest_framework/sentry_app.py

```diff
@@ -174,4 +162,26 @@ def validate(self, attrs):
         if not get_current_value("isInternal") and not get_current_value("author"):
             raise ValidationError({"author": "author required for public integrations"})
 
+        # validate globally unique name for public integrations, and org-level uniqueness for internal integrations
+        current_name = get_current_value("name")
+        current_is_internal = get_current_value("isInternal")
+        if current_name:
+            slug = generate_slug(
+                current_name,
+                is_internal=current_is_internal,
+                organization_slug=self.organization.slug,
+            )
+            if current_is_internal:
+                queryset = SentryApp.with_deleted.filter(owner=self.organization, slug=slug)
```

Comment:
@nolachen btw, this can still cause a bug if two orgs have the same first 6 characters. You will end up hitting the DB constraint later. I don't think we should put the owner as part of the filter.